File: src/main/java/com/dg/watcher/watching/retrieving/ApkRetrieving.java
package com.dg.watcher.watching.retrieving;

import hudson.FilePath;
import hudson.model.AbstractBuild;
import java.io.File;
import java.util.ArrayList;
import static com.dg.watcher.base.Const.APK_DEFAULT_DIR;
import static com.dg.watcher.base.Const.APK_EXTENSION;
import static java.io.File.separator;
import static org.apache.commons.io.FileUtils.listFiles;


public class ApkRetrieving {
    public static File retrieveApk(AbstractBuild<?, ?> build, String customApkDir) {
        ArrayList<File> apkFiles = retrieveApkFiles(createPathToApk(build, customApkDir));

        return !apkFiles.isEmpty() ? apkFiles.get(0) : null;
    }

    private static String createPathToApk(AbstractBuild<?, ?> build, String customApkDir) {
        String workspaceDir = retrieveWorkspaceDir(build);

        if(customDirSupplied(customApkDir)) {
            return workspaceDir + customApkDir;
        }
        else {
            return workspaceDir + APK_DEFAULT_DIR;
        }
    }

    private static String retrieveWorkspaceDir(AbstractBuild<?, ?> build) {
        FilePath workspaceFile = build.getWorkspace();

        return workspaceFile != null ? workspaceFile.getRemote() + separator : "";
    }

    private static boolean customDirSupplied(String customApkDir) {
        return !customApkDir.isEmpty();
    }

    private static ArrayList<File> retrieveApkFiles(String apkDirPath) {
        File apkDirectory = new File(apkDirPath);

        if(apkDirectory.exists()) {
            return new ArrayList<>(listFiles(apkDirectory, new String[] { APK_EXTENSION }, true));
        }

        return new ArrayList<>();
    }
}

File: src/main/java/com/dg/watcher/watching/retrieving/ApkRetrieving.kt
package com.dg.watcher.watching.retrieving

import com.dg.watcher.base.APK_DEFAULT_DIR
import com.dg.watcher.base.APK_EXTENSION
import hudson.model.AbstractBuild
import org.apache.commons.io.FileUtils.listFiles
import java.io.File
import java.io.File.separator
import java.util.*


fun retrieveApk(build: AbstractBuild<*, *>, customApkDir: String): File? {
    val apkFiles = retrieveApkFiles(createPathToApk(build, customApkDir))

    return if(!apkFiles.isEmpty()) apkFiles.first() else null
}

private fun createPathToApk(build: AbstractBuild<*, *>, customApkDir: String): String {
    val workspaceDir = retrieveWorkspaceDir(build)

    return if(customApkDir.isNotEmpty()) {
        workspaceDir + customApkDir
    }
    else {
        workspaceDir + APK_DEFAULT_DIR
    }
}

private fun retrieveWorkspaceDir(build: AbstractBuild<*, *>): String {
    val workspaceFile = build.getWorkspace()

    return if(workspaceFile != null) "${workspaceFile.remote}$separator" else ""
}

private fun retrieveApkFiles(apkDirPath: String): ArrayList<File> {
    val apkDirectory = File(apkDirPath)

    return if(apkDirectory.exists()) {
        ArrayList(listFiles(apkDirectory, arrayOf(APK_EXTENSION), true))
    }
    else {
        ArrayList()
    }
}